lib/3scale/backend/storage_helpers.rb

```diff
@@ -284,16 +280,35 @@ def cfg_unix_path_handler(options)
             options
           end
 
+          # Note that we can set reconnect_attempts to >= 0 because we use
+          # a monkey patch which implements a workaround for this issue
+          # that shows that there might be duplicated transactions when
+          # there's a timeout: https://github.com/redis/redis-rb/issues/668
+          # We should investigate if there are edge cases that can lead to
+          # duplicated commands because of this setting.
+          def cfg_reconnect_handler(options)
+            options[:reconnect_wait_seconds] = ThreeScale::Backend.listener? ? 0 : 5
+            options[:reconnect_attempts] = if ThreeScale::Backend.listener?
+                                             1
+                                           else
+                                             ThreeScale::Backend.configuration.redis.async ? Float::INFINITY : 1
+                                           end
```

Comment:
```suggestion
            options[:reconnect_attempts] = ThreeScale::Backend.listener? ? 1 : 2**30
```

Your original approach to make both sync and async wait a long time seemed good to me. So it is fine for me to set just a big value, not necessarily `INFINITY`.

But if you like to keep original sync behavior to `1` that's also fine. Everything looks good either way. Thank you!

Reply:
I tried your suggestion but it doesn't work because the sync mode doesn't accept a `:reconnect_wait_seconds` parameter. Instead it accepts an array like eg: `[1,5,15,30]`. This would mean: four attempts, wait 1 second after the first attempt; wait five seconds after the second attempt, etc. It also accepts a single number, but then it doesn't wait between attempts. If we pass it a big number like `2**30` it will enter loop that attempts to connect to redis `2**30` times with not wait time, and it takes all the resources in the box.

Reply:
At the end I came up with this https://github.com/3scale/apisonator/pull/412/commits/b2130c9cc7fb74e76f4ae0f34966c51f0c314ca3

This will make sync and async modes behave equal. Passing the sync client a `Hash(5)` behaves like an infinite array `arr = [5,5,5,...]`. No matter the index you send to it, e.g. `arr[1000]` or `arr[1000000]` it will always return `5` so the client is tricked to wait forever and wait 5 seconds between attempts.